### ISSUE TITLE
Example of how to wrap a concrete subclass within the abstract/final design pattern

### ISSUE THREAD

**Issue by BenjaminDAnjou** [patrick-kidger/equinox#1231]
I have a question regarding the [abstract/final pattern](https://docs.kidger.site/equinox/pattern/) that is recommended for `equinox`, specifically about this part:

> What about when you have an existing concrete class that you want to tweak just-a-little-bit? In this case, prefer composition over inheritance. Write a wrapper that forwards each method as appropriate. This is just as expressive, and means we keep these readable type-safe rules.

So for instance, consider the example class structure:

```python
from abc import abstractmethod
import jax.numpy as jnp
from jaxtyping import Float
import equinox as eqx

class AbstractShape(eqx.Module):

    area: eqx.AbstractVar[Float]

    @abstractmethod
    def compute_area(self):
        raise NotImplementedError
    
class Ellipse(AbstractShape):

    area: Float
    major_axis: Float
    minor_axis: Float

    def __init__(self, major_axis, minor_axis):
        self.major_axis = major_axis
        self.minor_axis = minor_axis
        self.area = self.compute_area()

    def compute_area(self):
        return jnp.pi * self.major_axis * self.minor_axis
    
# How to define a circle class according to the abstract/final pattern?
...
```

I would like to write a `Circle` class that is instantiated by passing the value of the `radius`. My _very strong immediate instinct_ is to write the `Circle` class as a subclass of the `Ellipse` class. But this is forbidden by the recommended pattern; you can never subclass a concrete class in the abstract/final design pattern.

One option is to write a `Circle` class that inherits directly from `AbstractShape`, and to rewrite the `compute_area` method of the `Circle` class. But if feels weird not to reuse the code of the `Ellipse` class. I guess I could just write a `compute_ellipse_area()` function elsewhere, and then call it with different inputs in the respective `__init__` of `Ellipse` and `Circle`. But it feels like it could get redundant quite quickly.

And then there is the wrapping approach mentioned in the doc. Should I `write a wrapper that forwards each method as appropriate` in that case, and how would that look like?

In summary, it would very helpful to me to see **how the abstract/final pattern should be implemented in the above example**. Or if I could be pointed to a similar example.

And if you think that's useful, maybe an example like this could be added to the doc, because I feel like it's a canonical example of subclassing.

**Comment by BenjaminDAnjou** [comment]
Ok I think other people helped me [figure out](https://stackoverflow.com/a/79943218/8800836) that this is what is meant:

```python
class Circle(AbstractShape):
    # 1. State the fields required by the AbstractShape interface
    area: Float
    
    # 2. State the unique fields for Circle
    radius: Float
    
    # 3. Store the underlying concrete instance for composition
    _ellipse: Ellipse

    def __init__(self, radius: Float):
        self.radius = radius
        
        # Instantiate the concrete class internally
        self._ellipse = Ellipse(major_axis=radius, minor_axis=radius)
        
        # Forward the property from the internal instance
        self.area = self._ellipse.area

    # 4. Forward methods to the internal instance
    def compute_area(self):
        return self._ellipse.compute_area()
```

It's counter-intuitive in the beginning but I'm already getting used to it.

**Comment by patrick-kidger** [comment]
Nice! I think for this case you have two options:

1. exactly as you're already doing, composition over inheritance. This is the most typical choice for problems of this sort, and in particular it grants the flexibility for `Circle` to offer a different API than `Ellipse`. (For example, it may remove `.{major,minor}_axis` and instead offer a `.radius`.)
 
2. the other option is to introduce a class constructor, either as a classmethod (typically if you own the `Ellipse` type):
    ```python
    class Ellipse(AbstractShape):
        major_axis: float
        minor_axis: float
    
        @classmethod
        def make_circle(cls, radius):
            return cls(radius, radius)
    ```
    or as a freestanding method (typically if you don't own the Ellipse type):
    ```python
    class Ellipse(AbstractShape):
        major_axis: float
        minor_axis: float
    
    def make_circle(radius):
        return Ellipse(radius, radius)
    ```

---

Two more unrelated notes on your code.

1. I would discourage `self.area = self.compute_area()`. Consider if you instantiate your `Ellipse` at the top level of your program, and then pass it cross a `jax.grad`, i.e. `jax.grad(some_fn)(Ellipse(...))` then you will get gradients with respect to `major_axis`, `minor_axis` and `area` separately, without regard for the way in which they are related to each other. It is better for each field to represent something independent of all the others. (And just call `self.compute_area()` wherever you need it in your program.)
2. the correct syntax for a scalar would be `jaxtyping.Float[jax.Array, ""]`, just a standalone `jaxtyping.Float` doesn't mean anything. (Because in this case jaxtyping doesn't know whether this should be a JAX array, a numpy array, a PyTorch tensor, a raw Python `float`, etc.)

I hope that helps!

**Comment by BenjaminDAnjou** [comment]
Yes that helps, thanks you. Regarding the following point:

> It is better for each field to represent something independent of all the others.

I'm guessing that this line in the second code snippet, the "composition-over-inheritance" one, solves this problem?

```
# Forward the property from the internal instance
self.area = self._ellipse.area
```

In that case, the fields from a tree that never intersects with itself, since `self._ellipse.area` is a refers to a completely different object.

Thanks for the `jaxtyping` tip. I just wanted to get the code to run, will delve more deeply into the docs later!

**Comment by patrick-kidger** [comment]
> I'm guessing that this line in the second code snippet, the "composition-over-inheritance" one, solves this problem?
> 
>> ```
>> # Forward the property from the internal instance
>> self.area = self._ellipse.area
>> ```

No, I don't think so. You end up with multiple independent arrays present throughout the tree: `self.area`and `self._ellipse.area` are not the same thing after a JAX transform, as per JAX's referential transparency. This is in addition to my original point that if done outside `jax.grad` then neither will have any relation to the minor/major axis they were created from.

**Comment by BenjaminDAnjou** [comment]
Ah right, my bad, what you're saying is completely unrelated to the `Circle` class. The problem is really just that the fields of the `Ellipse` class itself are not independent from each other (and if they were independent, there would be no issue in forwarding them to `Circle`).

So according to the discussion above, the following code removes `area` as a field and is A-OK:

```python
from abc import abstractmethod
import jax.numpy as jnp
from jaxtyping import Float, ArrayLike
import equinox as eqx

class AbstractShape(eqx.Module):

    @abstractmethod
    def compute_area(self):
        raise NotImplementedError
    
class Ellipse(AbstractShape):

    major_axis: Float[ArrayLike, ""]
    minor_axis: Float[ArrayLike, ""]

    def __init__(self, major_axis, minor_axis):
        self.major_axis = major_axis
        self.minor_axis = minor_axis

    def compute_area(self):
        return jnp.pi * self.major_axis * self.minor_axis


class Circle(AbstractShape):
    
    radius: Float[ArrayLike, ""]
    _ellipse: Ellipse

    def __init__(self, radius: Float):
        self.radius = radius
        self._ellipse = Ellipse(major_axis=radius, minor_axis=radius)

    def compute_area(self):
        return self._ellipse.compute_area()
```
